x-pack/plugins/uptime/public/hooks/use_composite_image.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { composeScreenshotRef } from '../lib/helper/compose_screenshot_images';
+import { ScreenshotRefImageData } from '../../common/runtime_types/ping/synthetics';
+
+/**
+ * Checks if two refs are the same. If the ref is unchanged, there's no need
+ * to run the expensive draw procedure.
+ */
+function isNewRef(a?: ScreenshotRefImageData, b?: ScreenshotRefImageData): boolean {
+  if (typeof a === 'undefined' || typeof b === 'undefined') return false;
+  const stepA = a.ref.screenshotRef.synthetics.step;
+  const stepB = b.ref.screenshotRef.synthetics.step;
+  return stepA.index !== stepB.index || stepA.name !== stepB.name;
+}
+
+/**
+ * Assembles the data for a composite image and returns the composite to a callback.
+ * @param imgRef the data and dimensions for the composite image.
+ * @param callback sends the composited image to this callback.
+ * @param url this is the composited image value, if it is truthy the function will skip the compositing process
+ */
+export const useCompositeImage = (
+  imgRef: ScreenshotRefImageData,
+  callback: React.Dispatch<string | undefined>,
+  url?: string
+): void => {
+  const [curRef, setCurRef] = React.useState<ScreenshotRefImageData | undefined>(undefined);
+
+  if (curRef === null) {
```

Comment:
Can `curRef` ever be null, since the type is `ScreenshotRefImageData | undefined`? I think you should be able to get rid of this logic with the above suggestion, but wondering what the intention is here.

Reply:
Is the intention that it always renders at least once with `curRef` undefined, to support the logic in `isNewRef`?

Reply:
Hey yes thanks for finding this, I wrote this code quickly to resolve a bug, and didn't look at it closely enough. Originally, the default value was `null`, but I later changed it to `undefined` because I wanted to use optional parameters for `isNewRef` and forgot to update this code.

You're 100% right on your recommendation above, and I think we don't even need an `undefined` default state here. This way we can remove any `undefined`/`null` checking. In the case of the first run, the `url` parameter is going to be `undefined`, so it doesn't matter that there's no new ref to compare against. Then in future runs, if the ref is unchanged, it won't re-render, otherwise it will render the new ref. I'll implement this and test it out.

Reply:
See https://github.com/elastic/kibana/pull/101678/commits/1e31984bd074d3631c12067e03189608419db6ae for implementation. This was a great catch I likely would've missed, thank you 💯